### DISCUSSION TITLE
Realism or variety of generated data?

### DISCUSSION THREAD

**Discussion by treylav** [lk-geimfari/mimesis#1292] [Q&A] [answered]
Hello.

I wanted to refactor the module with Russian locale and ran into an issues that are more related to the concept of a library than anything else. Let me explain with an example.

# Problem
Many Russian tax-related IDs — INN, KPP, OGRN — use the same code, which looks like `5300`, where `53` is the number of the region, `00` is the number of the specific tax office in this region that assigned concrete ID. List of all existing codes is [available online on the website of the Russian tax authority](https://www.nalog.gov.ru/html/sites/www.new.nalog.ru/SOUN/SOUN_DBF.ARJ).

If I want to implement the generation of such code in a library, I see several ways.

# Options
## Ignore format, use completely random codes
 For example, the author of the current [implementation of the OGRN](https://github.com/lk-geimfari/mimesis/blob/master/mimesis/builtins/ru.py#L162) generation chose this way.

The advantage of this approach:
  - maximum number of possible codes;
  - simplicity of the algorithm.
  
Disadvantages of this approach:
   - you may generate a code with a non-existent region: for example, there is no region with number 90 in Russia;
   - you may generate a code with a non-existent tax office.

## Use a set of predefined codes
For example, the author of the current [implementation of the KPP](https://github.com/lk-geimfari/mimesis/blob/master/mimesis/builtins/ru.py#L193) generation chose this way.

   - The advantage of this approach: 100% real data as source, generated codes becomes more realistic.
   - The disadvantage of this approach the number of generated codes is significantly reduced.

If you think that this is the best approach, we can improve the generation process using the official base mentioned above. [I dumped the codes from the database as a simple Python tuple](https://gist.github.com/treylav/11ed00a2b49d2ff68efe38321d8660f0). Unfortunately, I cannot suggest how best to store such an array (at the time of writing this post — 3409 entries), but together we can find a way to store this data for the purposes of the library.

## Use real region and random tax office number
This is the way I chose when I tried to start refactoring the module.

The advantages of this approach:
- realistic looking codes;
- still quite a large number of generated codes.

The disadvantage of this approach: it is still easy to recognize fake data by tax office indices, which are uncharacteristic for small regions.

# Conclusion
Personally, I'll be glad if we find a way to store an array of real codes from the database of the Russian tax authority.

The problem from the example may seem insignificant, but I would still like to understand, before contributing, how to balance the realism and variety of the generated data.

Thanks in advance.

**Comment by lk-geimfari** [comment, marked as answer]
I see no problem in creating the list of real codes. Feel free to create PR.
